src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -1275,6 +1283,35 @@ COOP_PINVOKE_HELPER(Object**, RhGetThreadStaticStorage, ())
     return pCurrentThread->GetThreadStaticStorage();
 }
 
+COOP_PINVOKE_HELPER(Object**, RhGetInlineThreadStaticStorage, ())
+{
+    Thread* pCurrentThread = ThreadStore::RawGetCurrentThread();
+    return pCurrentThread->GetInlineThreadStaticStorage();
+}
+
+NOINLINE
+static Object* InitAndGetInlinedThreadStaticBaseForType(uint32_t typeIndex)
+{
+    TypeManager* pTm = GetRuntimeInstance()->GetSingleTypeManager();
+    ASSERT(pTm != NULL);
+
+    return RhpGetInlinedThreadStaticBaseSlow(typeIndex, pTm);
```

Comment:
`RuntimeExport` methods can be only called from managed code. They cannot be called from C/C++ runtime code - it would result into non-walkable stack.

Reply:
If it is a tailcall, then there is no call. But to guarantee it is a tailcall this must be in assembly. And since the asm helper may need the type manager, it is simpler to just pass it in - as it was before. 

The on-demand allocation forced a few things to be reverted back to where they were. :-\

Reply:
The new helper looks like this (on Linux x64):
```asm
System.Collections.Tests`RhpGetThreadStaticBaseForType:
    0x555555477b3f <+0>:  push   rbx
->  0x555555477b40 <+1>:  push   r12
    0x555555477b42 <+3>:  mov    rbx, rdi
    0x555555477b45 <+6>:  mov    r12, rsi
    0x555555477b48 <+9>:  mov    rax, qword ptr fs:[0x0]
    0x555555477b54 <+21>: add    rax, -0xf0
    0x555555477b5b <+28>: mov    rax, qword ptr [rax + 0x98]
    0x555555477b62 <+35>: test   rax, rax
    0x555555477b65 <+38>: je     0x555555477b70            ; <+49>
    0x555555477b67 <+40>: mov    rax, qword ptr [rax + 8*r12 + 0x10]
    0x555555477b6c <+45>: pop    r12
    0x555555477b6e <+47>: pop    rbx
    0x555555477b6f <+48>: ret
    0x555555477b70 <+49>: mov    rdi, rbx
    0x555555477b73 <+52>: mov    rsi, r12
    0x555555477b76 <+55>: pop    r12
    0x555555477b78 <+57>: pop    rbx
    0x555555477b79 <+58>: jmp    0x5555557154e0            ; S_P_CoreLib_Internal_Runtime_ThreadStatics__GetInlinedThreadStaticBaseSlow at ThreadStatics.cs:64
```

Reply:
I think it still looks better than the original version, I wish the TLS access pattern was more restricted about what it might trash though.

Reply:
The new helper now looks like this (linux x64):

```asm
UnitTests`RhpGetInlinedThreadStaticBase:
    0x55555545c79f <+0>:  mov    rax, qword ptr fs:[0x0]
    0x55555545c7ab <+12>: add    rax, -0xf0

->  0x55555545c7b2 <+19>: mov    rax, qword ptr [rax + 0x98]
    0x55555545c7b9 <+26>: test   rax, rax
    0x55555545c7bc <+29>: je     0x555555503810            ; S_P_CoreLib_Internal_Runtime_ThreadStatics__GetInlinedThreadStaticBaseSlow at ThreadStatics.cs:39
    0x55555545c7c2 <+35>: ret
```

A lot nicer. This is roughly what JIT will inline.

Reply:
for reference, this is how it looks on win x64:
```asm
00007FF7A7503A30  mov         eax,dword ptr [_tls_index (07FF7A790E038h)]  
00007FF7A7503A36  mov         r8,qword ptr gs:[58h]  
00007FF7A7503A3F  mov         r8,qword ptr [r8+rax*8]  
00007FF7A7503A43  mov         eax,10h  
00007FF7A7503A48  add         rax,r8  

00007FF7A7503A4B  mov         rax,qword ptr [rax+98h]  
00007FF7A7503A52  test        rax,rax  
00007FF7A7503A55  je          S_P_CoreLib_Internal_Runtime_ThreadStatics__GetInlinedThreadStaticBaseSlow (07FF7A7649F20h)  
00007FF7A7503A5B  ret  
```